projects/RabbitMQ.Client/client/impl/Frame.cs

```diff
@@ -169,22 +164,22 @@ private InboundFrame(FrameType type, int channel, ReadOnlyMemory<byte> payload,
             _rentedArray = rentedArray;
         }
 
-        private static void ProcessProtocolHeader(Stream reader)
+        private static async ValueTask ProcessProtocolHeader(PipeReader reader, ReadOnlyMemory<byte> protocolError)
         {
             try
             {
-                byte b1 = (byte)reader.ReadByte();
-                byte b2 = (byte)reader.ReadByte();
-                byte b3 = (byte)reader.ReadByte();
+                byte b1 = protocolError.Span[0];
+                byte b2 = protocolError.Span[1];
+                byte b3 = protocolError.Span[2];
                 if (b1 != 'M' || b2 != 'Q' || b3 != 'P')
                 {
                     throw new MalformedFrameException("Invalid AMQP protocol header from server");
                 }
 
-                int transportHigh = reader.ReadByte();
-                int transportLow = reader.ReadByte();
-                int serverMajor = reader.ReadByte();
-                int serverMinor = reader.ReadByte();
+                int transportHigh = protocolError.Span[3];
+                int transportLow = protocolError.Span[4];
+                int serverMajor = protocolError.Span[5];
+                int serverMinor = await reader.ReadByteAsync().ConfigureAwait(false); ;
```

Comment:
this is bad you should read all bytes you need at minimum before calling ProcessProtocolHeader and make it sync

Reply:
This is an edge case that'll happen at most once during a connections lifetime, and it'll shut down right after, so I'm not worried about that here.